README
DBIx-LogAny version 0.01
====================

This module was created from DBIx::Log4perl but uses Log::Any instead.

DBIx-LogAny is an extension for DBI to selectively log SQL, parameters,
result-sets, transactions etc with Log::Any.

You will need the DBI, Log::Any, Log::Any::Adapter and a logging
module like Log::Log4perl.

Please see the NOTES section in the pod for incompatibilities.

Please make sure you have an up to date Test::More before running the
tests. You will need at least 0.62.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

NOTE: To to a full test you need a working DBI and DBD with access to
a database where you can create a table. Once you have this set up
you should run perl Makefile.PL like this:

perl Makefile.PL --dsn=dsn --user=dbuser --password=dbpassword --table=table

where:

  dsn is a valid DBI connection dsn e.g. dbd::mysql::database
  dbuser is a valid user to access database
  dbpassword is the password for dbuser
    (may be omitted if not password required)
  table is the name of a table to create for testing
    (NOTE: this table will be dropped if it already exists)

DEPENDENCIES

This module requires these other modules and libraries:

  DBI 1.50
  Log::LogAny 0.14
  Test::More 0.62

COPYRIGHT AND LICENCE

Copyright (C) 2012 by Martin J. Evans

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension DBIx::LogAny.

0.06 Friday May 19 2017

  [BUG FIXES]

  New test added in 0.05 showed a flaw with the table created which used
  a char where a varchar is required.

0.05 Friday May 19 2017

  [BUG FIXES]

  In LogAny.pm put the package definition at the start for correct imports
  thanks to Jens Rehsack.

  [ENHANCEMENTS]

  When a statement handle method is called with an already prepared
  statement handle log the SQL statement not the handle - thanks
  to Jens Rehsack.

  [DOCUMENTATION]

  Various fixes to documentation thanks to Jens Rehsack.

  [MISCELLANEOUS]

  Add travis setup thanks to Jens Resack.

0.04 Thursday December 9 2012

  [Bug Fixes]

  A misspelt Module:::Loaded, Module::Load

0.03 Thursday December 6 2012

  [Bug Fixes]

  Module::Loaded was missing from dependencies in Makefile.PL
  Thanks to CPAN smokers for spotting.

0.02 Friday November 30 2012

  [ENHANCEMENTS]

  Added dbix_la_category attribute so you can override the category passed
  to get_logger.

  Added dbix_la_logger attribute so you can pass in your own log handle.

  Added DBIX_LA_LOG_STORE so you can log calls to STORE seperately.

  [MISCELLANEOUS]

  Switched to use Log::Log4perl's register_wrapper instead of using
  caller_depth.

  Updated SYNOPSIS to include more/better examples.

0.01 Wednesday November 28 2012

  First version based on DBIx::Log4perl


t/lib.pl
# $Id: lib.pl 4157 2010-07-05 10:45:22Z martin $
use Cwd;
use File::Spec;
use Test::More;

my $logtmp1;
my $logtmp2;

sub get_config
{
    my @v;
    if (-f ($file = 't/dbixl4p.config')  ||
	-f ($file = '../dbixl4p.config') ||
	-f ($file = 'dbixl4p.config')) {
	open IN, "$file";
	while (<IN>) {
	    chomp;
	    if ($_ eq 'UNDEF') {
		push @v, undef;
	    } else {
		push @v, $_;
	    }
	}
    }
    return @v;
}

sub config
{
    my $td = File::Spec->tmpdir or die q/Can't find a temporary directory to use. Please set TMPDIR, TEMP or TMP environment variables to a valid cirectory/;

    $logtmp1 = File::Spec->catfile($td, 'dbixroot.log');
    note("log file 1 is $logtmp1");
    $logtmp2 = File::Spec->catfile($td, 'dbix.log');
    note("log file 2 is $logtmp2");
    return($logtmp1, $logtmp2);
}

sub check_log
{
    my ($s, $file) = @_;
    $$s = "";
    return 0 if (! -r $file);
    my ($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size, $atime,$mtime,$ctime,$blksize,$blocks) = stat($file);
    return 0 if ($size <= 0);
    open IN, "<$file";
    while (<IN>) {$$s .= $_};
    close IN;
    unlink $file;
    diag($$s) if ($ENV{TEST_VERBOSE});
    return 1;
}
1;


lib/DBIx/LogAny/db.pm
# $Id: db.pm 245 2006-07-25 14:20:59Z martin $
use strict;
use warnings;
use DBI;
use Data::Dumper;
use Module::Loaded;

package DBIx::LogAny::db;
use Log::Any;
@DBIx::LogAny::db::ISA = qw(DBI::db DBIx::LogAny);
use DBIx::LogAny::Constants qw (:masks $LogMask);

# $_glogger is not relied upon - it is just a fallback
my $_glogger;

my $_counter;                   # to hold sub to count

BEGIN {
    my $x = sub {
        my $start = shift;
        return sub {$start++}};
    $_counter = &$x(0);         # used to count dbh connections
}


sub STORE{
    my $dbh = shift;
    my @args = @_;

    my $h = $dbh->{private_DBIx_LogAny};
    # as we don't set private_DBIx_LogAny until the connect method sometimes
    # $h will not be set
    $dbh->_dbix_la_debug($h, 2, "STORE($h->{dbh_no})", @args)
        if ($h && ($h->{logmask} & DBIX_LA_LOG_INPUT));

    return $dbh->SUPER::STORE(@args);
}

sub get_info
{
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};
    my $value = $dbh->SUPER::get_info(@args);

    $dbh->_dbix_la_debug($h, 2, "get_info($h->{dbh_no})", @args, $value)
        if ($h->{logmask} & DBIX_LA_LOG_INPUT);
    return $value;
}
sub prepare {
    my($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};
    my $ctr = $h->{new_stmt_no}(); # get a new unique stmt counter in this dbh
    if (($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL)) &&
            (caller !~ /^DBIx::LogAny/o) &&
                (caller !~ /^DBD::/o)) { # e.g. from selectall_arrayref
        $dbh->_dbix_la_debug($h, 2, "prepare($h->{dbh_no}.$ctr)", $args[0]);
    }

    my $sth = $dbh->SUPER::prepare(@args);
    if ($sth) {
        $sth->{private_DBIx_LogAny} = $h;
        $sth->{private_DBIx_st_no} = $ctr;
    }

    return $sth;
}

sub prepare_cached {
    my($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};
    my $ctr = $h->{new_stmt_no}();
    if (($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL)) &&
            (caller !~ /^DBIx::LogAny/o) &&
                (caller !~ /^DBD::/o)) { # e.g. from selectall_arrayref
        $dbh->_dbix_la_debug($h, 2,
                              "prepare_cached($h->{dbh_no}.$ctr)", $args[0]);
    }

    my $sth = $dbh->SUPER::prepare_cached(@args);
    if ($sth) {
        $sth->{private_DBIx_LogAny} = $h;
        $sth->{private_DBIx_st_no} = $ctr;
    }
    return $sth;
}

sub do {
    my ($dbh, @args) = @_;
    my $h = $dbh->{private_DBIx_LogAny};

    $h->{Statement} = $args[0];
    $dbh->_dbix_la_debug($h, 2, "do($h->{dbh_no})", @args)
        if ($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL));

    my $affected = $dbh->SUPER::do(@args);

    if (!defined($affected)) {
        $dbh->_dbix_la_error(2, 'do error for ', @args)
            if (($h->{logmask} & DBIX_LA_LOG_ERRCAPTURE) &&
                    !($h->{logmask} & DBIX_LA_LOG_INPUT)); # not already logged
    } elsif (defined($affected) && $affected eq '0E0' &&
                 ($h->{logmask} & DBIX_LA_LOG_WARNINGS)) {
        $dbh->_dbix_la_warning(2, 'no effect from ', @args);
    } elsif (($affected ne '0E0') && ($h->{logmask} & DBIX_LA_LOG_INPUT)) {
        $dbh->_dbix_la_debug($h, 2, "affected($h->{dbh_no})", $affected);
        $dbh->_dbix_la_debug($h, 2, "\t" . $dbh->SUPER::errstr)
            if (!defined($affected));
    }
    return $affected;
}

sub selectrow_array {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};

    if ($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL)) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_la_debug($h,
                2,
                "selectrow_array($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_la_debug($h, 2,
                                  "selectrow_array($h->{dbh_no})", @args);
        }
    }

    if (wantarray) {
	my @ret = $dbh->SUPER::selectrow_array(@args);
	$dbh->_dbix_la_debug($h, 2, "result($h->{dbh_no})", @ret)
	  if ($h->{logmask} & DBIX_LA_LOG_OUTPUT);
	return @ret;

    } else {
	my $ret = $dbh->SUPER::selectrow_array(@args);
	$dbh->_dbix_la_debug($h, 2, "result($h->{dbh_no})", $ret)
	  if ($h->{logmask} & DBIX_LA_LOG_OUTPUT);
	return $ret;
    }
}

sub selectrow_arrayref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};

    if ($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL)) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_la_debug(
                $h, 2,
                "selectrow_arrayref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_la_debug(
                $h, 2, "selectrow_arrayref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectrow_arrayref(@args);
    $dbh->_dbix_la_debug($h, 2, "result($h->{dbh_no})", $ref)
      if ($h->{logmask} & DBIX_LA_LOG_OUTPUT);
    return $ref;
}

sub selectrow_hashref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};

    if ($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL)) {
        if ((scalar(@args) > 0) && (ref $args[0])){
            $dbh->_dbix_la_debug(
                $h, 2,
                "selectrow_hashref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args)
        } else {
            $dbh->_dbix_la_debug($h, 2,
                                  "selectrow_hashref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectrow_hashref(@args);
    # no need to show result - fetch will do this
    return $ref;

}

sub selectall_arrayref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};
    if ($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL)) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_la_debug(
                $h, 2,
                "selectall_arrayref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_la_debug(
                $h, 2, "selectall_arrayref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectall_arrayref(@args);
    $dbh->_dbix_la_debug($h, 2, "result($h->{dbh_no})", $ref)
      if ($h->{logmask} & DBIX_LA_LOG_OUTPUT);
    return $ref;
}

sub selectall_hashref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};
    if ($h->{logmask} & (DBIX_LA_LOG_INPUT|DBIX_LA_LOG_SQL)) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_la_debug(
                $h, 2,
                "selectall_hashref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_la_debug($h, 2,
                                  "selectall_hashref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectall_hashref(@args);
    # no need to show result - fetch will do this
    return $ref;

}

sub _make_counter {
    my $start = shift;
    return sub {$start++}
};

sub connected {

    my ($dbh, $dsn, $user, $pass, $attr) = @_;

    my %h = ();
    $h{dbh_no} = &$_counter();
    $h{new_stmt_no} = _make_counter(0); # get a new stmt count for this dbh

	# if passed a Log4perl log handle use that
	if (exists($attr->{dbix_la_logger})) {
	    $h{logger} = $attr->{dbix_la_logger};
	} elsif (exists($attr->{dbix_la_category})) {
	    $h{category} = $attr->{dbix_la_category};
	    $h{logger} = Log::Any->get_logger(category => $h{category});
	} else {
        $h{logger} = Log::Any->get_logger(category => __PACKAGE__);
    }

    # save log mask
    $h{logmask} = $attr->{dbix_la_logmask} if (exists($attr->{dbix_la_logmask}));
    # save error regexp
    $h{err_regexp} = $attr->{dbix_la_ignore_err_regexp}
        if (exists($attr->{dbix_la_ignore_err_regexp}));

    # take global log mask if non defined
    $h{logmask} = $LogMask unless (exists($h{logmask}));

    $_glogger = $h{logger};

    $h{ll_loaded} = Module::Loaded::is_loaded('Log::Log4perl');

    # make sure you don't change the depth before calling get_logger:
    local $Log::Log4perl::caller_depth = $Log::Log4perl::caller_depth + 4
        if $h{ll_loaded};

    $h{dbd_specific} = 0;
    $h{driver} = $dbh->{Driver}->{Name};

    $dbh->{private_DBIx_LogAny} = \%h;

    if ($h{logmask} & DBIX_LA_LOG_CONNECT) {
        local $Data::Dumper::Indent = 0;
        $h{logger}->debug(
            "connect($h{dbh_no}): " .
                (defined($dsn) ? $dsn : '') . ', ' .
                    (defined($user) ? $user : '') . ', ' .
                        Data::Dumper->Dump([$attr], [qw(attr)]))
            if $h{logger}->is_debug;
        no strict 'refs';
        my $v = "DBD::" . $dbh->{Driver}->{Name} . "::VERSION";
        $h{logger}->info("DBI: " . $DBI::VERSION,
                         ", DBIx::LogAny: " . $DBIx::LogAny::VERSION .
                             ", Driver: " . $h{driver} . "(" .
                                 $$v . ")")
            if $h{logger}->is_info;
    }

    #
    # If capturing errors then save any error handler and set_err Handler
    # passed to us and replace with our own.
    #
    if ($h{logmask} & DBIX_LA_LOG_ERRCAPTURE) {
        $h{HandleError} = $attr->{HandleError}
            if (exists($attr->{HandleError}));
        $h{HandleSetErr} = $attr->{HandleSetErr}
            if (exists($attr->{HandleSetErr}));
        $dbh->{HandleError} = \&_error_handler;
        $dbh->{HandleSetErr} = \&_set_err_handler;
    }
    return;

}
sub clone {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_LogAny};
    if ($h->{logmask} & DBIX_LA_LOG_CONNECT) {
        $dbh->_dbix_la_debug($h, 2, "clone($h->{dbh_no})", @args);
    }

    return $dbh->SUPER::clone(@args);
}

sub disconnect {
    my $dbh = shift;

    if ($dbh) {
	my $h;
	eval {
	    # Avoid
	    # (in cleanup) Can't call method "FETCH" on an undefined value
	    $h = $dbh->{private_DBIx_LogAny};
	};
	if (!$@ && $h && defined($h->{logger})) {
            if ($h->{logmask} & DBIX_LA_LOG_CONNECT) {
                local $Log::Log4perl::caller_depth =
                    $Log::Log4perl::caller_depth + 2
                        if $h->{ll_loaded};
                $dbh->_dbix_la_debug($h, 2, "disconnect($h->{dbh_no})");
            }
	}
    }
    return $dbh->SUPER::disconnect;

}

sub begin_work {
    my $dbh = shift;
    my $h = $dbh->{private_DBIx_LogAny};

    $dbh->_dbix_la_debug($h, 2, "start transaction($h->{dbh_no})")
        if ($h->{logmask} & DBIX_LA_LOG_TXN);

    return $dbh->SUPER::begin_work;
}

sub rollback {
    my $dbh = shift;
    my $h = $dbh->{private_DBIx_LogAny};

    $dbh->_dbix_la_debug($h, 2, "roll back($h->{dbh_no})")
        if ($h->{logmask} & DBIX_LA_LOG_TXN);

    return $dbh->SUPER::rollback;
}

sub commit {
    my $dbh = shift;

    my $h = $dbh->{private_DBIx_LogAny};
    $dbh->_dbix_la_debug($h, 2, "commit($h->{dbh_no})")
        if ($h->{logmask} & DBIX_LA_LOG_TXN);

    return $dbh->SUPER::commit;
}

sub last_insert_id {
    my ($dbh, @args) = @_;
    my $h = $dbh->{private_DBIx_LogAny};

    $dbh->_dbix_la_debug(
        $h, 2, Data::Dumper->Dump([\@args], ["last_insert_id($h->{dbh_no})"]))
        if ($h->{logmask} & DBIX_LA_LOG_INPUT);

    my $ret = $dbh->SUPER::last_insert_id(@args);
    $dbh->_dbix_la_debug($h, 2, "\t" . DBI::neat($ret))
      if ($h->{logmask} & DBIX_LA_LOG_INPUT);
    return $ret;
}


#
# Error handler to capture errors and log them
# Whatever, errors are passed on.
# if the user of DBIx::LogAny passed in an error handler that is called
# before returning.
#
sub _error_handler {
    my ($msg, $handle, $method_ret) = @_;

    my $dbh = $handle;
    my $lh;
    my $h = $handle->{private_DBIx_LogAny};
    my $out = '';

    $lh = $_glogger;
    $lh = $h->{logger} if ($h && exists($h->{logger}));
    return 0 if (!$lh);

    if (!$lh->is_fatal) {
        goto FINISH;
    }

    if ($h && exists($h->{err_regexp})) {
        if ($dbh->err =~ $h->{err_regexp}) {
            goto FINISH;
        }
    }
    # start with error message, state and err
    $out .=  '  ' . '=' x 60 . "\n  $msg\n";
    $out .= "err() = " . $handle->err . "\n";
    $out .= "state() = " . $handle->state . "\n";

    if ($DBI::lasth) {
        $out .= "  lasth type: $DBI::lasth->{Type}\n"
            if ($DBI::lasth->{Type});
        $out .= "  lasth Statement ($DBI::lasth):\n    " .
            "$DBI::lasth->{Statement}\n"
                if ($DBI::lasth->{Statement});
    }
    # get db handle if we have an st
    my $type = $handle->{Type};
    my $sql;
    if ($type eq 'st') {        # given statement handle
        $dbh = $handle->{Database};
        $sql = $handle->{Statement};
    } else {
        # given db handle
        # We've got other stmts under this db but we'll deal with those later
        $sql = 'Possible SQL: ';
        $sql .= "/$h->{Statement}/" if (exists($h->{Statement}));
        $sql .= "/$dbh->{Statement}/"
            if ($dbh->{Statement} &&
                    (exists($h->{Statement}) &&
                         ($dbh->{Statement} ne $h->{Statement})));
    }

    my $dbname = exists($dbh->{Name}) ? $dbh->{Name} : "";
    my $username = exists($dbh->{Username}) ? $dbh->{Username} : "";
    $out .= "  DB: $dbname, Username: $username\n";
    $out .= "  handle type: $type\n  SQL: " . DBI::neat($sql) . "\n";
    $out .= '  db Kids=' . $dbh->{Kids} .
        ', ActiveKids=' . $dbh->{ActiveKids} . "\n";
    $out .= "  DB errstr: " . $handle->errstr . "\n"
        if ($handle->errstr && ($handle->errstr ne $msg));

    if (exists($h->{ParamValues}) && $h->{ParamValues}) {
        $out .= "  ParamValues captured in HandleSetErr:\n    ";
        foreach (sort keys %{$h->{ParamValues}}) {
            $out .= "$_=" . DBI::neat($h->{ParamValues}->{$_}) . ",";
        }
        $out .= "\n";
    }
    if ($type eq 'st') {
        my $str = "";
        if ($handle->{ParamValues}) {
            foreach (sort keys %{$handle->{ParamValues}}) {
                $str .= "$_=" . DBI::neat($handle->{ParamValues}->{$_}) . ",";
            }
        }
        $out .= "  ParamValues: $str\n";
        $out .= "  " .
            Data::Dumper->Dump([$handle->{ParamArrays}], ['ParamArrays'])
                  if ($handle->{ParamArrays});
    }
    my @substmts;
    # get list of statements under the db
    push @substmts, $_ for (grep { defined } @{$dbh->{ChildHandles}});
    $out .= "  " . scalar(@substmts) . " sub statements:\n";
    if (scalar(@substmts)) {
        foreach my $stmt (@substmts) {
            $out .= "  stmt($stmt):\n";
            $out .= '    SQL(' . $stmt->{Statement} . ")\n  "
                if ($stmt->{Statement} &&
                        (exists($h->{Statement}) &&
                             ($h->{Statement} ne $stmt->{Statement})));
            if (exists($stmt->{ParamValues}) && $stmt->{ParamValues}) {
                $out .= '   Params(';
                foreach (sort keys %{$stmt->{ParamValues}}) {
                    $out .= "$_=" . DBI::neat($stmt->{ParamValues}->{$_}) . ",";
                }
                $out .= ")\n";
            }
        }
    }

    if (exists($dbh->{Callbacks})) {
        $out .= "  Callbacks exist for " .
            join(",", keys(%{$dbh->{Callbacks}})) . "\n";
    }
    local $Carp::MaxArgLen = 256;
    $out .= "  " .Carp::longmess("DBI error trap");
    $out .= "  " . "=" x 60 . "\n";

    local $Log::Log4perl::caller_depth = $Log::Log4perl::caller_depth + 1
        if $h->{ll_loaded};

    $lh->fatal($out);

  FINISH:
    if ($h && exists($h->{ErrorHandler})) {
        return $h->{ErrorHandler}($msg, $handle, $method_ret);
    } else {
        return 0;               # pass error on
    }
}

#
# set_err handler so we can capture ParamValues before a statement
# is destroyed.
# If the use of DBIx::LogAny passed in an error handler that is
# called before returning.
#
sub _set_err_handler {
    my ($handle, $err, $errstr, $state, $method) = @_;

    # Capture ParamValues
    if ($handle) {
	my $h = $handle->{private_DBIx_LogAny};
	$h->{ParamValues} = $handle->{ParamValues}
	    if (exists($handle->{ParamValues}));
	return $h->{HandleSetErr}(@_) if (exists($h->{HandleSetErr}));
    }
    return 0;
}


1;


lib/DBIx/LogAny/Constants.pm
use strict;
use warnings;

package DBIx::LogAny::Constants;
require Exporter;
our @ISA = qw(Exporter);

use constant DBIX_LA_LOG_DEFAULT => 0xdd;
use constant DBIX_LA_LOG_ALL => 0xffffff;
use constant DBIX_LA_LOG_INPUT => 0x1;
use constant DBIX_LA_LOG_OUTPUT => 0x2;
use constant DBIX_LA_LOG_CONNECT => 0x4;
use constant DBIX_LA_LOG_TXN => 0x8;
use constant DBIX_LA_LOG_ERRCAPTURE => 0x10;
use constant DBIX_LA_LOG_WARNINGS => 0x20;
use constant DBIX_LA_LOG_ERRORS => 0x40;
use constant DBIX_LA_LOG_DBDSPECIFIC => 0x80;
use constant DBIX_LA_LOG_DELAYBINDPARAM => 0x100;
use constant DBIX_LA_LOG_SQL => 0x200;

our $LogMask = DBIX_LA_LOG_DEFAULT;

our @EXPORT = ();
our @EXPORT_OK = qw ($LogMask);
our @EXPORT_MASKS = qw(DBIX_LA_LOG_DEFAULT
		       DBIX_LA_LOG_ALL
		       DBIX_LA_LOG_INPUT
		       DBIX_LA_LOG_OUTPUT
		       DBIX_LA_LOG_CONNECT
		       DBIX_LA_LOG_TXN
		       DBIX_LA_LOG_ERRCAPTURE
		       DBIX_LA_LOG_WARNINGS
		       DBIX_LA_LOG_ERRORS
		       DBIX_LA_LOG_DBDSPECIFIC
               DBIX_LA_LOG_DELAYBINDPARAM
               DBIX_LA_LOG_SQL
		     );
our %EXPORT_TAGS = (masks => \@EXPORT_MASKS);
Exporter::export_ok_tags('masks');


1;


Makefile.PL
use 5.008000;
use ExtUtils::MakeMaker;
use Getopt::Long();
use Data::Dumper;

use vars qw($opt);
$opt = {};

Getopt::Long::GetOptions(
    $opt,
    "help" => sub {Usage()},
    "dsn=s",
    "user=s",
    "password:s",
    "table=s",
    ) || die Usage();


my ($dsn, $user, $password, $table) = ('', '', 'UNDEF', '');

$dsn = $ENV{DBI_DSN} if (exists($ENV{DBI_DSN}));
$user = $ENV{DBI_USER} if (exists($ENV{DBI_USER}));
$password = $ENV{DBI_PASS} if (exists($ENV{DBI_PASS}));
$table = $ENV{DBI_TABLE} if (exists($ENV{DBI_TABLE}));

$dsn = $opt->{dsn} if (exists($opt->{dsn}));
$user = $opt->{user} if (exists($opt->{user}));
$password = $opt->{password} if (exists($opt->{password}));
$table = $opt->{table} if (exists($opt->{table}));

$dsn ||= "dbi:DBM:";
$table ||= 'drop_me_dbixl4p';

open OUT, ">dbixl4p.config";
print OUT "$dsn\n";
print OUT "$user\n";
print OUT "$password\n";
print OUT "$table\n";
close OUT;

warn "\nPlease NOTE\nYou need to set and export the DBI_* environment variables or supply suitable\nvalues on the command line (see Perl Makefile.PL --help) to run a complete test\n\n";

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'DBIx::LogAny',
    VERSION_FROM      => 'lib/DBIx/LogAny.pm', # finds $VERSION
    LICENSE           => 'perl',
    PREREQ_PM         => {
        Log::Any => 0.14,
        Log::Any::Adapter => 0.10,
        DBI => 1.50,
        File::Spec => 0,
        Data::Dumper => 0,
        Test::More => 0.62,
        Module::Loaded => 0,
	Scalar::Util => 0},
    META_MERGE => {
        resources => {
            bugtracker => 'http://rt.cpan.org/Public/Dist/Display.html?Name=DBIx-LogAny',
            repository => 'https://github.com/mjegh/DBIx-LogAny.git',
            Mailinglist => 'http://dbi.perl.org/support',
            homepage => 'http://search.cpan.org/dist/DBIx-LogAny',
        }
    },
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/DBIx/LogAny.pm', # retrieve abstract from module
       AUTHOR         => 'Martin J. Evans <mjevans@cpan.org>') : ()),
);


sub Usage {
  print STDERR <<"USAGE";
Usage: perl $0 [options]

Possible options are:

  --dsn=<db>         The DBI DSN to pass to DBI->connect
  --user=<user>      Use the username <user> for running the test suite;
                     defaults to no username
  --password=<pwd>   Use the password <pwd> for running the test suite;
                     defaults to no password
  --table=<table>    A table to be used as a test table in the DB
                     defaults to dbixl4p
USAGE
  exit 1;
}
